tests/stack/test_api.py

```diff
@@ -529,3 +578,84 @@ def test_download_resource(self, stack_api):
+            if resource_service == api.WORKSPACE_SERVICE:
+                if workspace_api.NOTEBOOK == \
+                        resource_properties[api.WORKSPACE_RESOURCE_OBJECT_TYPE]:
+                    os.makedirs(os.path.dirname(resource_properties[
+                                                api.WORKSPACE_RESOURCE_SOURCE_PATH]))
+                    with open(resource_properties[api.WORKSPACE_RESOURCE_SOURCE_PATH], 'w') as f:
+                        f.write("print('test')\n")
+                if resource_properties[api.WORKSPACE_RESOURCE_OBJECT_TYPE] == \
+                        workspace_api.DIRECTORY:
+                    os.makedirs(resource_properties[api.WORKSPACE_RESOURCE_SOURCE_PATH])
+            elif resource_service == api.DBFS_SERVICE:
+                if resource_properties[api.DBFS_RESOURCE_IS_DIR]:
+                    os.makedirs(resource_properties[api.DBFS_RESOURCE_SOURCE_PATH])
+                else:
+                    with open(resource_properties[api.DBFS_RESOURCE_SOURCE_PATH], 'w') as f:
+                        f.write("print('test')\n")
+
+        # Run deploy command on stack and then verify stack status. Validation for the stack
+        # status is done within the deploy command, so if the function doesn't error, then
+        # the generated stack is valid.
+        new_stack_status_1 = stack_api.deploy_config(test_stack)
+        # Asserting some correctness criteria for the generated stack status.
+        assert new_stack_status_1.get(api.STACK_RESOURCES) == test_stack.get(api.STACK_RESOURCES)
+        for deployed_resource in new_stack_status_1.get(api.STACK_DEPLOYED):
+            # All functions to pull the deployed output were mocked to return deploy_output
+            assert deployed_resource.get(api.RESOURCE_DEPLOY_OUTPUT) == test_deploy_output
+
+        # Test that when passing in a  status that the new status generated will still be valid.
+        new_stack_status_2 = stack_api.deploy_config(test_stack, stack_status=TEST_STATUS)
+        # Test some extra correctness criteria for the generated stack status.
+        assert new_stack_status_2.get(api.STACK_RESOURCES) == test_stack.get(api.STACK_RESOURCES)
+        for deployed_resource in new_stack_status_2.get(api.STACK_DEPLOYED):
+            assert deployed_resource.get(api.RESOURCE_DEPLOY_OUTPUT) == test_deploy_output
+
+    def test_download(self, stack_api):
+        """
+            stack_api.download_from_config should should call _download_resource on all of its
+            resources but only call _download_workspace on the workspace resources.
+        """
+        stack_api._download_workspace = mock.MagicMock()
+        stack_api._download_resource = mock.Mock(wraps=stack_api._download_resource)
+
+        stack_api.download_from_config(TEST_STACK)
+        # because there is only two workspace resources, _download
+        assert stack_api._download_resource.call_count == 5
```

Comment:
Could you explain why 5? and why 2? You can create some simpler config just for testing this method.